lib/drab/client.ex

```diff
@@ -148,64 +148,86 @@ defmodule Drab.Client do
   @spec api_version() :: String.t()
   def api_version(), do: @client_lib_version
 
+  @spec get_controller_module(Plug.Conn.t()) :: atom() | nil
+  defp get_controller_module(conn) do
+    try do
+      Phoenix.Controller.controller_module(conn)
```

Comment:
For both this and the one in `get_controller_action_name` is there no way to more safely check this without incurring an exception?  Exceptions are pretty cheap when not thrown (although setting up the exception handler is not free in any case), but they are much less cheap when they are actually thrown, which would be very common in LiveView cases.  If not this looks good as it is.  :-)

Reply:
I don't like it either use exceptions in a library but I haven't found a better way, both [`Phoenix.Controller.controller_module/1`](https://hexdocs.pm/phoenix/Phoenix.Controller.html#controller_module/1) and [`Phoenix.Controller.action_name/1`](https://hexdocs.pm/phoenix/Phoenix.Controller.html#action_name/1) raise if the Controller is unavailable.

Any suggestions for some alternatives  of curse are welcome.

Reply:
Hmm, all it checks is `conn.private[:phoenix_controller]`, similar for the other, which is technically internal and not exposed, although it's never been changed.  I wonder if direct access might be warranted, thoughts?

Reply:
In fact my first version was similar, it was a check against `: live_view_module` in the assigns ( `if !conn.assigns[:live_view_module] do`), but it was a solution too specific for LV, so searching for a  a general solution I thought that is better to use the official API because, although unlikely, the parameters for a direct access could be change tomorrow.

Reply:
Actually I don't know how much heavy thrown exceptions are, it could be interesting to setup some benchmark so to be aware of the actual costs. In any case we could assume that we don't want exception management in Drab and use the `conn.private[:phoenix_controller]` hack, and switch to the API based solutions if tomorrow things changes.

Reply:
Hmm, let's test this case:
```
╰─➤  mix bench map_lookup_or_throw                                                                                                1 ↵
Operating System: Linux"
CPU Information: AMD Phenom(tm) II X6 1090T Processor
Number of Available Cores: 6
Available memory: 15.67 GB
Elixir 1.8.1
Erlang 21.2.2

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 2 s
memory time: 2 s
parallel: 1
inputs: large-10000-existing, large-9999-not-existing, medium-10-existing, medium-9-not-existing, small-1-not-existing, small-2-existing
Estimated total run time: 3 min


Benchmarking Access with input large-10000-existing...
Benchmarking Access with input large-9999-not-existing...
Benchmarking Access with input medium-10-existing...
Benchmarking Access with input medium-9-not-existing...
Benchmarking Access with input small-1-not-existing...
Benchmarking Access with input small-2-existing...
Benchmarking Map.fetch with input large-10000-existing...
Benchmarking Map.fetch with input large-9999-not-existing...
Benchmarking Map.fetch with input medium-10-existing...
Benchmarking Map.fetch with input medium-9-not-existing...
Benchmarking Map.fetch with input small-1-not-existing...
Benchmarking Map.fetch with input small-2-existing...
Benchmarking case with input large-10000-existing...
Benchmarking case with input large-9999-not-existing...
Benchmarking case with input medium-10-existing...
Benchmarking case with input medium-9-not-existing...
Benchmarking case with input small-1-not-existing...
Benchmarking case with input small-2-existing...
Benchmarking catch-Map.fetch! with input large-10000-existing...
Benchmarking catch-Map.fetch! with input large-9999-not-existing...
Benchmarking catch-Map.fetch! with input medium-10-existing...
Benchmarking catch-Map.fetch! with input medium-9-not-existing...
Benchmarking catch-Map.fetch! with input small-1-not-existing...
Benchmarking catch-Map.fetch! with input small-2-existing...
Benchmarking catch-direct-always-fail with input large-10000-existing...
Benchmarking catch-direct-always-fail with input large-9999-not-existing...
Benchmarking catch-direct-always-fail with input medium-10-existing...
Benchmarking catch-direct-always-fail with input medium-9-not-existing...
Benchmarking catch-direct-always-fail with input small-1-not-existing...
Benchmarking catch-direct-always-fail with input small-2-existing...

##### With input large-10000-existing #####
Name                               ips        average  deviation         median         99th %
case                           11.18 M      0.0895 μs     ±6.43%      0.0880 μs       0.106 μs
catch-Map.fetch!                9.60 M       0.104 μs   ±202.32%       0.100 μs       0.150 μs
Map.fetch                       9.54 M       0.105 μs   ±211.47%       0.100 μs        0.22 μs
Access                          7.39 M       0.135 μs    ±17.79%       0.130 μs        0.20 μs
catch-direct-always-fail        5.90 M       0.169 μs    ±56.58%       0.160 μs        0.25 μs

Comparison: 
case                           11.18 M
catch-Map.fetch!                9.60 M - 1.16x slower
Map.fetch                       9.54 M - 1.17x slower
Access                          7.39 M - 1.51x slower
catch-direct-always-fail        5.90 M - 1.89x slower

Memory usage statistics:

Name                        Memory usage
case                           529.25 KB
catch-Map.fetch!               529.25 KB - 1.00x memory usage
Map.fetch                      529.27 KB - 1.00x memory usage
Access                         529.25 KB - 1.00x memory usage
catch-direct-always-fail       529.28 KB - 1.00x memory usage

**All measurements for memory usage were the same**

##### With input large-9999-not-existing #####
Name                               ips        average  deviation         median         99th %
case                           13.22 M      0.0756 μs    ±15.54%      0.0700 μs       0.120 μs
Map.fetch                      10.46 M      0.0956 μs    ±16.55%      0.0900 μs       0.140 μs
Access                          8.05 M       0.124 μs    ±17.15%       0.120 μs       0.180 μs
catch-Map.fetch!                5.75 M       0.174 μs   ±228.92%       0.160 μs        0.27 μs
catch-direct-always-fail        5.69 M       0.176 μs   ±276.39%       0.160 μs        0.37 μs

Comparison: 
case                           13.22 M
Map.fetch                      10.46 M - 1.26x slower
Access                          8.05 M - 1.64x slower
catch-Map.fetch!                5.75 M - 2.30x slower
catch-direct-always-fail        5.69 M - 2.32x slower

Memory usage statistics:

Name                        Memory usage
case                           529.20 KB
Map.fetch                      529.20 KB - 1.00x memory usage
Access                         529.20 KB - 1.00x memory usage
catch-Map.fetch!                0.180 KB - 0.00x memory usage
catch-direct-always-fail       529.23 KB - 1.00x memory usage

**All measurements for memory usage were the same**


##### With input medium-10-existing #####
Name                               ips        average  deviation         median         99th %
case                           16.57 M      0.0604 μs     ±6.70%      0.0590 μs      0.0733 μs
catch-Map.fetch!               11.85 M      0.0844 μs     ±6.72%      0.0820 μs       0.105 μs
Map.fetch                      11.00 M      0.0909 μs   ±434.54%      0.0800 μs        0.21 μs
Access                         10.08 M      0.0992 μs     ±7.51%      0.0960 μs       0.124 μs
catch-direct-always-fail        6.17 M       0.162 μs   ±232.69%       0.150 μs        0.32 μs

Comparison: 
case                           16.57 M
catch-Map.fetch!               11.85 M - 1.40x slower
Map.fetch                      11.00 M - 1.51x slower
Access                         10.08 M - 1.64x slower
catch-direct-always-fail        6.17 M - 2.69x slower

Memory usage statistics:

Name                        Memory usage
case                               592 B
catch-Map.fetch!                   592 B - 1.00x memory usage
Map.fetch                          616 B - 1.04x memory usage
Access                             592 B - 1.00x memory usage
catch-direct-always-fail           624 B - 1.05x memory usage

**All measurements for memory usage were the same**

##### With input medium-9-not-existing #####
Name                               ips        average  deviation         median         99th %
case                           18.67 M      0.0536 μs    ±19.03%      0.0500 μs      0.0800 μs
Map.fetch                      12.84 M      0.0779 μs     ±6.68%      0.0760 μs      0.0970 μs
Access                         10.50 M      0.0952 μs    ±27.39%      0.0900 μs       0.140 μs
catch-Map.fetch!                6.55 M       0.153 μs    ±77.30%       0.140 μs        0.25 μs
catch-direct-always-fail        6.27 M       0.159 μs   ±207.46%       0.150 μs        0.23 μs

Comparison: 
case                           18.67 M
Map.fetch                      12.84 M - 1.45x slower
Access                         10.50 M - 1.78x slower
catch-Map.fetch!                6.55 M - 2.85x slower
catch-direct-always-fail        6.27 M - 2.98x slower

Memory usage statistics:

Name                        Memory usage
case                               552 B
Map.fetch                          552 B - 1.00x memory usage
Access                             552 B - 1.00x memory usage
catch-Map.fetch!                   184 B - 0.33x memory usage
catch-direct-always-fail           584 B - 1.06x memory usage

**All measurements for memory usage were the same**

##### With input small-1-not-existing #####
Name                               ips        average  deviation         median         99th %
case                           16.93 M      0.0591 μs     ±8.03%      0.0590 μs      0.0830 μs
Map.fetch                      14.18 M      0.0705 μs    ±41.16%      0.0680 μs      0.0890 μs
Access                         11.45 M      0.0874 μs    ±26.97%      0.0800 μs       0.130 μs
catch-Map.fetch!                6.78 M       0.148 μs   ±182.52%       0.140 μs        0.26 μs
catch-direct-always-fail        6.20 M       0.161 μs   ±288.91%       0.150 μs        0.26 μs

Comparison: 
case                           16.93 M
Map.fetch                      14.18 M - 1.19x slower
Access                         11.45 M - 1.48x slower
catch-Map.fetch!                6.78 M - 2.50x slower
catch-direct-always-fail        6.20 M - 2.73x slower

Memory usage statistics:

Name                        Memory usage
case                               232 B
Map.fetch                          232 B - 1.00x memory usage
Access                             232 B - 1.00x memory usage
catch-Map.fetch!                   184 B - 0.79x memory usage
catch-direct-always-fail           264 B - 1.14x memory usage

**All measurements for memory usage were the same**

##### With input small-2-existing #####
Name                               ips        average  deviation         median         99th %
case                           17.26 M      0.0579 μs    ±48.24%      0.0580 μs      0.0810 μs
catch-Map.fetch!               13.85 M      0.0722 μs    ±55.45%      0.0700 μs       0.100 μs
Map.fetch                      13.08 M      0.0765 μs     ±9.96%      0.0720 μs       0.109 μs
Access                         11.07 M      0.0903 μs     ±8.15%      0.0890 μs       0.110 μs
catch-direct-always-fail        6.18 M       0.162 μs   ±220.42%       0.150 μs        0.35 μs

Comparison: 
case                           17.26 M
catch-Map.fetch!               13.85 M - 1.25x slower
Map.fetch                      13.08 M - 1.32x slower
Access                         11.07 M - 1.56x slower
catch-direct-always-fail        6.18 M - 2.79x slower

Memory usage statistics:

Name                        Memory usage
case                               272 B
catch-Map.fetch!                   272 B - 1.00x memory usage
Map.fetch                          296 B - 1.09x memory usage
Access                             272 B - 1.00x memory usage
catch-direct-always-fail           304 B - 1.12x memory usage

**All measurements for memory usage were the same**
```

So in summary, matching via a case extraction is always fastest.  The `catch-direct-always-fail` will be the case where Phoenix's functions will lookup and fail (and it always fails for the same reason that phoenix hard-codes in its keys).  The `Access` pattern is the `some_map[some_key]` lookup, it's not great either.

In short, a case matcher would be about double the speed on average.  I'm not sure that's really worth caring about then.  I'm voting to accept it with the Phoenix calls as it is then.  Agree?

Reply:
Thank you for the benchmark, interesting results.
I agree, differences are not substantial, and in case we can change our mind anytime